### PR TITLE
Fixed thread race condition when updating PartsPanel

### PR DESCRIPTION
**Pull request by david-hall**
* Ensure that propertyChangeEvents are handled on the EventDispatchThread
* Cleanup several trivial compiler warnings.

I got distracted from testing the ColorMapping persistence to a file. Now that this PartsPanel bug is fixed, I'll try to finish testing the ColorMapping stuff soon.

### CODE REVIEW

**Comment by david-hall** [comment]
I just added a second commit which improves several minor issues on the PartsPanel. Specifically, the sorting, tool tips, displaying panels and totals... Details are in the commit description. The first commit is the bug fix for the thread race condition and should definitely be incorporated. The second commit that I added tonight is pretty much new features so take a look and see if you like it. BTW, I think it also makes it easier to modify the data we pass to the panel and how the PartsPanel orders and groups it in case we ever get around to adding to it down the road... Now back to testing the Color Mapping persistence...

File: src/main/java/org/vorthmann/zome/ui/PartsPanel.java [hunk 1]
```diff
@@ -113,130 +113,98 @@ public Object getValueAt( int row, int col )
...
-            String name = evt .getPropertyName();
-            if ( name .equals( "addBall" ) )
-            {
-                ++ balls;
-                this .fireTableCellUpdated( 0, 0 );
+            PropertyChangeHandler handler = new PropertyChangeHandler(evt .getPropertyName());
+            if (SwingUtilities.isEventDispatchThread()) {
+                handler.run();
+            } else {
+                SwingUtilities.invokeLater(handler);
```
Comment (vorth) [inline src/main/java/org/vorthmann/zome/ui/PartsPanel.java]: Seems that this is only necessary if there is something on the wrong thread triggering this propertyChange, and that's the problem to be fixed.  Do you know where the trigger is?

**Comment by vorth** [comment]
I took a look at the fix for the race condition, and if I understand correctly, it involves SwingUtilities.invokelater().  While that is a useful and important tool, it is only necessary when running on some thread other than the event processing thread, or if one wants to use the event queue architecturally, to serialized cascading events or something.

So the question is, why is a property change being triggered on some other thread?  I’d prefer to solve the problem that way.

I’m also concerned about the addition of PartInfo, but possibly for the wrong reason.  Years ago, I intentionally made the UI code dependent only on the generic notion of a Controller.  This was a change designed to enable use of some UI technology other than Swing, even something non-Java-based.  The string-based event channel would be easy to bridge to any native libraries, or other languages like Javascript.  (At the time, I was looking at Cocoa, Apple’s native UI toolkit.)

That decoupling is less important to me know, I’ll admit, since I’m more interested in leaving Java behind entirely.  I’ll also admit that shoving data through the string-based channel is very limiting, as you’ve seen.

In the end, I think I’m OK with PartInfo, but I wanted to explain the context, so you have more info to go on the next time you see a similar need.

**Comment by david-hall** [comment]
I figured that making the event handler thread-safe was a better design than being concerned with what threads might change the property and thus fire the event. In any case, it should be pretty easy to identify what thread(s) is/are calling it. I'll dig a little more in that area and let you know what I discover.

**Comment by david-hall** [comment]
Looks like the ExclusiveAction class is specifically designed for running on the non-UI thread. If an ExclusiveAction ultimately fires a PropertyChangeEvent such as manifestationAdded, then that handler is not executed on the UI thread. Specifically, I discovered that creating a strut by dragging with the mouse works on the UI thread, but hiding a strut with Ctrl-H is not processed on the UI thread. Creating a strut with JoinBalls is not on the UI thread either. I suspect that most, if not all of the classes derived from ChangeManifestations may not be run on the UI thread and could potentially have similar considerations, but that's just speculation at this point  I ran out of time and didn't dig much deeper at this time. It looks like there are only a couple of places in DocumentFrame where the ExclusiveAction constructor is called so it should be easy to figure out what's happening. 

I don't think I'll get to look at it much more until Monday or Tuesday, but I wanted to pass along what I found so far. Hopefully this much will stir up something that will help you determine which approach we should take to the multiple thread issue. Let me know if there's something else in particular you want me to look into.

**Comment by vorth** [comment]
Thanks for digging, David.  Yes, this was helpful.  Yes, most significant edits use ExclusiveAction, to block out other user edits while still freeing up the event handing thread (EHT).  I had forgotten this detail.  Your approach to making the PropertyChangeListener thread safe is correct, though I’ll come back to this question at the end of the email.

The issue does make me question my architecture.  It makes me want to segregate PropertyChangeEvents (PCEs) by source.  One kind of PCE represents the UI reacting to a UI gesture, with no round-trip to the controller and model layers.  For example, often a PCE will fire from the EHT, say, during a slider drag, checkbox click, or some other control gesture.  I was thinking about this pattern of PCE firing and handling when I reacted against using invokeLater; it feels like an anti-pattern when the PCE fires on the EHT in the first place.

Another source of PCEs is the model layer, in response to an UndoableEdit, as you’ve seen for the parts panel.  From memory, I'm not sure how many of these there are.

Finally, there is quite a bit of state in the controllers themselves, state that is not saved in the file.  Changes to controller state will trigger PCEs also.  Controller state is probably a bad thing in many (but not all) cases… it represents “models” that are not explicitly called out as such.

It would be very interesting to do a survey of all PCE firing points, and see if a classification emerges, and possibly some patterns for correct implementation.

Getting back to the original race condition: the race is between the ExclusiveAction thread firing a PCE and something else happening on the EHT.  What is the "something else"?  I'll review the original commit here to see if I can understand better.  I'm thinking along these lines: if we know we're firing a PCE from the ExclusiveAction background thread, perhaps we should use a Runnable to *fire* the event, then all PCLs can continue to assume that they are running on the EHT.  Perhaps another way to say it: any PCE fired from the model layer should use a Runnable to fire.

I look forward to your thoughts.

**Comment by vorth** [comment]
I see your excellent comment (again) in the code, explaining the race.  I think this confirms my conclusion above: any PCE known to emit from the background thread should be emitted via invokeLater somehow.  I'm still thinking about how to do that without putting a SwingUtilities dependency in the model layer.  I need to look at the details of how this particular PCE is fired.

**Pull request merged**